modules/app/main.tf

```diff
@@ -0,0 +1,57 @@
+data "aws_ami" "ami_app" {
+  owners = ["099720109477"] # Canonical
+  filter {
+    name   = "image-id"
+    values = [var.app_ami]
+  }
+}
+
+resource "aws_instance" "ec2_app" {
+  ami = data.aws_ami.ami_app.id
+  instance_type = var.app_instance_type
+  subnet_id = var.app_subnet
+  vpc_security_group_ids = [aws_security_group.sg_app.id]
+  key_name = var.app_provisioning_key
+  associate_public_ip_address = var.app_associate_public_ip_address
+  user_data = templatefile("${path.module}/templates/user_data.tpl", { mongo_address = var.app_mongo_address})
+  root_block_device {
+    delete_on_termination = true
+  }
+  tags = var.app_tags
+}
+
+resource "aws_security_group" "sg_app" {
+  name = "App Security Group"
+  description = "the app sg"
+  vpc_id = var.app_vpc_id
+}
+
+resource "aws_security_group_rule" "sg_rule_allow_http" {
+  description = "Allow tcp traffic on port 8080"
+  cidr_blocks = var.app_ingress_cidr_blocks
+  from_port = 8080
+  protocol = "tcp"
+  security_group_id = aws_security_group.sg_app.id
+  to_port = 8080
+  type = "ingress"
+}
+
+resource "aws_security_group_rule" "sg_rule_allow_ssh" {
```

Comment:
allow 22 to be able to get to the `app` box, also it could be useful as a jumpbox to the mongo EC2 instance to troubleshoot.

Reply:
to reduce the attack surface, makes sense to only whitelist the certain known IP addresses. Though we use only key-based SSH auth, still `0.0.0.0` is a way too wide.